lib/puppet/node.rb

```diff
@@ -89,7 +89,7 @@ def environment=(env)
     unless @environment.nil?
       # always set the environment parameter. It becomes top scope $environment for a manifest during catalog compilation.
       @parameters[ENVIRONMENT] = @environment.name.to_s
-      self.environment_name = @environment.name if instance_variable_defined?(:@environment_name)
+      self.environment_name = @environment.name
```

Comment:
Not sure why the `instance_variable_defined?` check was there in the first place

Reply:
I was initially concerned that if we always set the `environment_name`, then it might alter how the `environment` reader returns the environment
https://github.com/puppetlabs/puppet/blob/c3ddb9552685dc45b96ad045453b0f38b49fbed3/lib/puppet/node.rb#L60-L75

But I think it's ok because we only reach https://github.com/puppetlabs/puppet/blob/c3ddb9552685dc45b96ad045453b0f38b49fbed3/lib/puppet/node.rb#L92 after checking if `@environment` is not nil https://github.com/puppetlabs/puppet/blob/c3ddb9552685dc45b96ad045453b0f38b49fbed3/lib/puppet/node.rb#L89

So if later we ask for the `Node#environment` it'll always return `@environment` https://github.com/puppetlabs/puppet/blob/c3ddb9552685dc45b96ad045453b0f38b49fbed3/lib/puppet/node.rb#L60-L62

And that's the only place where the `Node` reads the `environment_name` instance variable, so I think always keeping `environment.name` and `environment_name` insync is safe.

That's not true for `parameters[ENVIRONMENT]` though, e.g. https://github.com/puppetlabs/puppet/blob/c3ddb9552685dc45b96ad045453b0f38b49fbed3/lib/puppet/node.rb#L150

So I'm :+1: on this.